File: src/euler_014.jl
function collatz(n, c = 0)
    (n == 1) ? c : (n % 2 == 0) ? collatz(div(n, 2), c + 1) : collatz(3n + 1, c + 1)
end

function answer(lim)
    map(collatz, 1:lim) |> argmax
end

answer(1_000_000) |> print

File: src/euler_014.hs
import Data.List

collatz :: Int -> Maybe (Int,Int)
collatz 1 = Nothing
collatz n
  | even n = Just (n, div n 2)
  | odd  n = Just (n, 3 * n +1)

awnser = last . sortOn (length . unfoldr collatz)

main = putStrLn $ show $ awnser [1..1000000]
